### ISSUE TITLE
no-unreachable not reporting all unreachable code

### ISSUE THREAD

**Issue by jmccaskill** [eslint/eslint#3939]
1. The version of ESLint you are using.
v1.5.1

2. What you did.
An `if-else` where the `if` and the `else` both have a `return`, so anything after the `if-else` is unreachable.
```
/* eslint no-unreachable: 2 */
'use strict';

function testCase(param) {
    if (param) {
        return true;
        console.log('this is unreachable code that is reported as an error.');
    }
    else {
        return false;
        console.log('this is also unreachable code that is reported as an error.');
    }

    console.error('this is unreachable code that is not reported as an error.');
}
```

3. What you expected to happen.
I expected the code after the `if-else` to be reported as unreachable.

4. What actually happened.
The code after the `if-else` was not reported as unreachable.
My original test case just had the `console.error` at the bottom of the function.  Just to make sure I didn't configure something incorrectly, I added the `console.log`s after the `return`s in the `if` and `else`, and those were both correctly reported as unreachable.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
Code path analysis would be able to help with this.

**Comment by nzakas** [comment]
Blocked on code path analysis. Cc @mysticatea

**Comment by mysticatea** [comment]
Yeah, I'm rewriting totally no-unreachable in #3559.
It should fix this.
